numpy/_core/src/multiarray/arraytypes.c.src

```diff
@@ -359,7 +359,17 @@ NPY_NO_EXPORT int
     @type@ temp;  /* ensures alignment */
 
 #if @is_int@
-    if (PyLong_Check(op)) {
+    int arg_converted = 0;
+    int is_pylong = PyLong_Check(op);
+    if (!(is_pylong || PyArray_Check(op))) {
```

Comment:
The `PyArray_Check` here is a little weird but if I don't include it that breaks some tests related to masked arrays. In particular, the default integer mask fill value, 99999, is outside of the range of 8 and 16 bit integers so if you have a masked structured array with (say) an `i1` field, setting the default fill value raises an `OverflowError`. Yuck!

Didn't want to deal with that, so added a carve-out for array-likes. I could also check e.g. subtypes of `PyUnicode_Type` and only convert strings, but it seems better to do this overflow checking on anything that can be converted to a pyint.

Reply:
... I was struggling with that quite a bit with NEP 50 too (and it probably may well still crop up in some places).  Don't love the full exceptions for arrays, but I agree with the general approach.

Unless anyone else has a thought: let's just add a comment why we have this exception here?